samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c

```diff
@@ -0,0 +1,824 @@
+				->sin_addr.s_addr;
+			broker4->sin_family = AF_INET;
+			broker4->sin_port = htons(ctx.port);
+
+			inet_ntop(AF_INET, &broker4->sin_addr.s_addr,
+				  ipv4_addr, sizeof(ipv4_addr));
+			LOG_DBG("IPv4 Address found %s", ipv4_addr);
+
+			break;
+		}
+		LOG_DBG("ai_addrlen = %u should be %u or %u",
+			(unsigned int)addr->ai_addrlen,
+			(unsigned int)sizeof(struct sockaddr_in),
+			(unsigned int)sizeof(struct sockaddr_in6));
+		addr = addr->ai_next;
+		break;
+	}
+
+	/* Free the address. */
+	freeaddrinfo(result);
+}
+
+/**@brief Initialize the MQTT client structure
+ */
+static void client_init(void)
+{
+	/* Init MQTT client */
+	mqtt_client_init(&client);
+
+	/* Init MQTT broker */
+	broker_init();
+
+	/* MQTT client configuration */
+	client.broker = &broker;
+	client.evt_cb = mqtt_evt_handler;
+	client.client_id.utf8 = ctx.cid;
+	client.client_id.size = strlen(ctx.cid);
+	client.password = NULL;
+	client.user_name = NULL;
+	client.protocol_version = MQTT_VERSION_3_1_1;
+
+	/* MQTT buffers configuration */
+	client.rx_buf = rx_buffer;
+	client.rx_buf_size = sizeof(rx_buffer);
+	client.tx_buf = tx_buffer;
+	client.tx_buf_size = sizeof(tx_buffer);
+
+	/* MQTT transport configuration */
+	if (ctx.sec_transport == true) {
+		struct mqtt_sec_config *tls_config;
+
+		tls_config = &(client.transport).tls.config;
+		tls_config->peer_verify = TLS_PEER_VERIFY_REQUIRED;
+		tls_config->cipher_list = NULL;
+		tls_config->cipher_count = 0;
+		tls_config->sec_tag_count = 1;
+		tls_config->sec_tag_list = (int *)&ctx.sec_tag;
+		tls_config->hostname = ctx.url;
+		client.transport.type = MQTT_TRANSPORT_SECURE;
+	} else {
+		client.transport.type = MQTT_TRANSPORT_NON_SECURE;
+	}
+}
+
+/**@brief Initialize the file descriptor structure used by poll.
+ */
+static int fds_init(struct mqtt_client *c)
+{
+	if (c->transport.type == MQTT_TRANSPORT_NON_SECURE) {
+		fds.fd = c->transport.tcp.sock;
+	} else {
+#if defined(CONFIG_MQTT_LIB_TLS)
+		fds.fd = c->transport.tls.sock;
+#else
+		return -ENOTSUP;
+#endif
+	}
+
+	fds.events = POLLIN;
+
+	return 0;
+}
+
+static int do_mqtt_connect(void)
+{
+	int err = -EINVAL;
+
+	if (ctx.connected) {
+		return -EINPROGRESS;
+	}
+
+	client_init();
+
+	/* Start MQTT listening thread */
+	mqtt_thread_id = k_thread_create(&mqtt_thread,
+			mqtt_thread_stack,
+			K_THREAD_STACK_SIZEOF(mqtt_thread_stack),
+			mqtt_thread_fn, NULL, NULL, NULL,
+			THREAD_PRIORITY, 0, K_NO_WAIT);
+
+	err = mqtt_connect(&client);
+	if (err != 0) {
+		LOG_ERR("ERROR: mqtt_connect %d", err);
+		return err;
+	}
+
+	err = fds_init(&client);
+	if (err != 0) {
+		LOG_ERR("ERROR: fds_init %d", err);
+		return err;
+	}
+
+	/** start polling now for CONNACK */
+	if (err == 0) {
+		ctx.connected = true;
+		k_sem_give(&mqtt_connected);
+	}
+
+	return 0;
+}
+
+static int do_mqtt_disconnect(void)
+{
+	int err;
+
+	err = mqtt_disconnect(&client);
+	if (err) {
+		LOG_ERR("ERROR: mqtt_disconnect %d", err);
+	}
+	slm_at_mqtt_uninit();
+
+	return err;
```

Comment:
I guess that you need to suspend the thread here.
In connection, check the thread_id, if already created, do resume.
This is the key point o support re-connect to Broker

Reply:
The thread will bo blocked on a semaphore waiting for another connection, thus making it inactive. No need to suspend explicitly.